src/data/set/ncard.lean

```diff
@@ -0,0 +1,724 @@
+/-
+Copyright (c) 2023 Peter Nelson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Peter Nelson
+-/
+import data.finite.card
+import algebra.big_operators.finprod
+
+/-!
+# Noncomputable Set Cardinality
+
+We define the cardinality `set.ncard s` of a set `s` as a natural number. This function is
+noncomputable (being defined in terms of `nat.card`) and takes the value `0` if `s` is infinite.
+
+This can be seen as an API for `nat.card α` in the special case where `α` is a subtype arising from
+a set. It is intended as an alternative to `finset.card` and `fintype.card`,  both of which contain
+data in their definition that can cause awkwardness when using `set.to_finset`.  Using `set.ncard`
+allows cardinality computations to avoid `finset`/`fintype` completely, staying in `set` and letting
+finiteness be handled explicitly, or (where a `finite α` instance is present and the sets are
+in `set α`) via `auto_param`s.
+
+## Main Definitions
+
+* `set.ncard s` is the cardinality of the set `s` as a natural number, provided `s` is finite.
+  If `s` is infinite, then `set.ncard s = 0`.
+* `to_finite_tac` is a tactic that tries to synthesize an `set.finite s` argument with
+  `set.to_finite`. This will work for `s : set α` where there is a `finite α` instance.
+
+## Implementation Notes
+
+The lemmas in this file are very similar to those in `data.finset.card`, but with `set` operations
+instead of `finset`; most of the proofs invoke their `finset` analogues. Nearly all the lemmas
+require finiteness of one or more of their arguments. We provide this assumption with an
+`auto_param` argument of the form `(hs : s.finite . to_finite_tac)`, where `to_finite_tac` will find
+a `finite s` term in the cases where `s` is a set in a `finite` type.
+
+Often, where there are two set arguments `s` and `t`, the finiteness of one follows from the other
+in the context of the lemma, in which case we only include the ones that are needed, and derive the
+other inside the proof. A few of the lemmas, such as `ncard_union_le` do not require finiteness
+arguments; they are are true by coincidence due to junk values.
+-/
+
+open_locale classical
```

Comment:
It's better nowadays to avoid `open_locale classical`. Add `[decidable_eq α]` and/or `[decidable (a ∈ s)]` in the lemmas statement that complain, and maybe `classical` at the beginning of certain proofs.

Reply:
Done!

Reply:
The first thing to try is to add `classical` in the proof, if the statement it's OK there is no need to add `[decidable_eq ...]` in the statement (I am sorry, my first message was confusing). The linter will complain. You can check you did the right thing by adding `#lint` at the end of the file.

Reply:
For example
```
lemma ncard_union_eq [decidable_eq α] (h : disjoint s t) (hs : s.finite . to_finite_tac)
  (ht : t.finite . to_finite_tac) :
  (s ∪ t).ncard = s.ncard + t.ncard :=
begin
  rw [ncard_eq_to_finset_card _ hs, ncard_eq_to_finset_card _ ht,
    ncard_eq_to_finset_card _ (hs.union ht),finite.to_finset_union],
  refine finset.card_union_eq _,
  rwa [finite.disjoint_to_finset],
end
```
Should be
```
lemma ncard_union_eq (h : disjoint s t) (hs : s.finite . to_finite_tac)
  (ht : t.finite . to_finite_tac) :
  (s ∪ t).ncard = s.ncard + t.ncard :=
begin
  classical,
  rw [ncard_eq_to_finset_card _ hs, ncard_eq_to_finset_card _ ht,
    ncard_eq_to_finset_card _ (hs.union ht),finite.to_finset_union],
  refine finset.card_union_eq _,
  rwa [finite.disjoint_to_finset],
end
```

Reply:
Done!

Reply:
Sorry, I misread your first comment, and didn't see the second. I think the last push does what you were talking about. There are now only two `decidable` instance parameters, both pertaining to a lemma statement involving `ite`.